runners/core-java/src/main/java/org/apache/beam/runners/core/UnboundedReadFromBoundedSource.java

```diff
@@ -118,6 +118,11 @@ public void populateDisplayData(DisplayData.Builder builder) {
   public static class BoundedToUnboundedSourceAdapter<T>
       extends UnboundedSource<T, BoundedToUnboundedSourceAdapter.Checkpoint<T>> {
 
+    @SuppressWarnings("unused") // for Kryo
+    private BoundedToUnboundedSourceAdapter() {
+      this.boundedSource = null;
+    }
+
     private BoundedSource<T> boundedSource;
```

Comment:
Unfortunately this prevents safe object publishing since this isn't final and can't be changed to be final.

Please add test to show that deserialization via Kryo works.